src/rules.jl

```diff
@@ -1,260 +1,313 @@
-export rule_f!, rule_h!, rule_i1!, rule_i2!, rule_pi!, rule_b!, rule_c!
+export Rule_f, Rule_h, Rule_i1, Rule_i2, Rule_pi, Rule_c, Rule_b, Match
+export match_rule, rewrite!
+# export rule_f!, rule_h!, rule_i1!, rule_i2!, rule_pi!, rule_b!, rule_c!
 
-function check_rule_f(zxd::ZXDiagram{T,U,P}, v1::T, v2::T) where {T<:Integer, U<:Integer, P}
-    msg = true
-    if v1 <= nv(zxd) && v2 <= nv(zxd)
-        if !has_edge(zxd.g, v1, v2)
-            msg = "Spiders $(v1) and $(v2) are not connected!"
-        end
-        if spider_type(zxd, v1) == spider_type(zxd, v2)
-            if ~(spider_type(zxd, v1) == Z || spider_type(zxd, v1) == X)
-                msg = "Spiders $(v1) and $(v2) are not X or Z spiders!"
+abstract type AbstractRule end
+
+struct Rule_f <: AbstractRule end
```

Comment:
I think we can use only one stuck `BasicRule{L}` instead of declaring a lot of similar things here. Then you can use `BasicRule{:h}` as your traits in the `match` and `rewrite!` function